driver/scsi_operation.c

```diff
@@ -548,25 +548,60 @@ WnbdModeSense(_In_ PWNBD_DISK_DEVICE Device,
 }
 
 UCHAR
-WnbdReportLuns(_In_ PSCSI_REQUEST_BLOCK Srb) {
+WnbdReportLuns(_In_ PWNBD_EXTENSION DeviceExtension,
+    _In_ PSCSI_REQUEST_BLOCK Srb)
+{
+    WNBD_LOG_DEBUG(": Enter");
+    ASSERT(DeviceExtension);
     ASSERT(Srb);
 
+    UCHAR SrbStatus = SRB_STATUS_SUCCESS;
     PVOID DataBuffer = SrbGetDataBuffer(Srb);
     ULONG DataTransferLength = SrbGetDataTransferLength(Srb);
 
     if (NULL == DataBuffer) {
-        return SRB_STATUS_INTERNAL_ERROR;
+        SrbStatus = SRB_STATUS_INTERNAL_ERROR;
+        goto Exit;
+    }
+
+    if (DataTransferLength < sizeof(LUN_LIST)) {
+        WNBD_LOG_ERROR("Buffer of size %d insufficient to store "
+            "LUN list of size: %d", DataTransferLength, sizeof(LUN_LIST));
+        SrbStatus = SRB_STATUS_BAD_SRB_BLOCK_LENGTH;
+        goto Exit;
     }
 
-    // the LUN address is expected to be zero
     RtlZeroMemory(DataBuffer, DataTransferLength);
+    PLUN_LIST LunList = (PLUN_LIST) DataBuffer;
+    
+    ULONG LunCount = RtlNumberOfSetBits(&DeviceExtension->ScsiBitMapHeader);
```

Comment:
Aren't we suppose to retrieve just the luns that are associated with a given target? This will go over *all* luns.